src/main/java/com/es/phoneshop/model/product/history/HttpSessionHistoryService.java

```diff
@@ -0,0 +1,58 @@
+package com.es.phoneshop.model.product.history;
+
+import com.es.phoneshop.model.product.ArrayListProductDao;
+import com.es.phoneshop.model.product.Product;
+import com.es.phoneshop.model.product.exceptions.ProductNotFoundException;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpSession;
+import java.util.ArrayList;
+import java.util.List;
+
+public class HttpSessionHistoryService implements HistoryService {
+    protected static final String HISTORY_SESSION_ATTRIBUTE = HttpSessionHistoryService.class + ".history";
+    private static final int MAX_HISTORY_SIZE = 3;
+
+    public static HttpSessionHistoryService getInstance() {
+        return SingletonHolder.instance;
+    }
+
+    public static class SingletonHolder {
+        public static final HttpSessionHistoryService instance = new HttpSessionHistoryService();
+    }
+
+    private void add(List<Product> historyProducts, Long productId) throws ProductNotFoundException {
+        Product product = ArrayListProductDao.getInstance().getProduct(productId);
+        int i;
+        for (i = 0; i < historyProducts.size(); i++) {
```

Comment:
Currently this whole functionality works incorrectly, because I can see the first viewed item always (even if I visit 10 other item pdps after). So, reimplement and simplify this method as much as possible.